Design/Fit functions in python.md

```diff
@@ -0,0 +1,198 @@
+```
+Composite wrappers will also define the `fn` attributes (where `n` is a number) for example: `c.f0`, `c.f1`, ... The following three lines will have the same effect:
+```
+  c.f1.f0.A0 = 2
+  c[1][0]['A0'] = 2
+  c['f1.f0.A0'] = 2
+```
+
+The wrappers will override `__iadd__(self, func)` and `__delitem__(self, i)` (`i` is an `int`, `func` is a `FitFunctionWrapper`) to implement adding and deleting members via `+=` and `del` operators:
+```
+  spectrum += Lorentzian(PeakCentre=3)
+  del spectrum[0]
+```
+
+Implement `__len__(self)` to return the number of member functions.
+```
+  n_peaks = len(spectrum)
+```
+
+### Setting ties
+
+Ties are set with the `tie` method. Both keyword arguments and dictionaries of name/value pairs can be used as arguments.
+```
+  func.tie(A0=2.0)
+  func.tie({'f1.A2': '2*f0.A1', 'f2.A2': '3*f0.A1 + 1'})
+```
+
+Tying to a constant value that a parameter currently has can also be done with the `fix` method.
+```
+  func.fix('A0')
+  func.fix('f1.A2', 'f2.A2')
+```
+
+Ties are removed with the `removeTie` or `free` methods.
+```
+  func.free('f1.Sigma', 'f3.Sigma')
+  func.removeTie('f1.Sigma', 'f3.Sigma')
+```
+
+Composite functions should be able to set ties on all members with a single call:
+```
+  c.tieAll('Sigma')
+```
+The above call should tie all `Sigma` parameters in the member functions. It is equivalent to calling
+```
+  c.tie({'f1.Sigma': 'f0.Sigma', 'f2.Sigma': 'f0.Sigma', 'f3.Sigma': 'f0.Sigma', ...})
+```
+Similarly a call to `fixAll` fixes all parameters with the given name to their current value.
+```
+  spectrum.fixAll('FWHM')
+  spectrum.fixAll('f1.FWHM': 0.01)
+```
+Calling `fixAll()`/`freeAll()` without arguments fixes/frees all parameters of a function.
+
+### Setting constraints
+
+Constraints are set with `constraints` method passing constraints as strings:
+```
+  g.constraints('-1 < PeakCentre < 1', 'Sigma > 0')
+```
+
+For composite functions constraints can be set both through the composite function and its members:
+```
+  c = LinearBackground() + Gaussian(PeakCentre=1) + Gaussian(PeakCentre=2)
+  c.constraints('f1.Sigma > 0')
+  c[2].constraints('0 < Sigma < 1')
+```
+
+Composite functions should be able to set constraints on all members with a single call:
+```
+  c.constrainAll('Height > 1', 'Sigma > 0')
+```
+This should constrain the named parameters in all members that have them. Members that don't have these parameters should be skipped.
+
+## Fitting
+
+Python function `Fit` should accept instances of `FitFunctionWrapper` as its `Function` argument. `Fit` does't modify the input function but the returned value has the output function with the fitted parameters. The namedtuple returned by `Fit` must contain the fitted function:
+```
+  res = Fit(func, ws)
+  output_func = res.Function
+```
+
+## Function evaluation
+
+`FitFunctionWrapper` class will implement `__call__(...)` method to evaluate the function. The method will be able to accept a variety of input types:
```

Comment:
regarding plotting, having a python method is more intuitive to me than an external function.
```
f=Lorentzian()
f.plot(*args, **args) # workspace, workspaceindex, startX, endX
```
Sometimes the workspace and workspaceIndex have been already passed with the SetWorkspace method.
```
f=TeixeiraWaterSQE()
f.setMatrixWorkspace(workspace, workspaceIndex) # determines attribute Q
f.plot(startX, endX)
```